### PR TITLE
#28 #37 Fix a couple bugs for byte array zeroing and failure handling if the IronWeb frame can't be loaded

### PR DESCRIPTION
**Pull request by ernieturner**
#28 Clone byte arrays on input before passing to the frame to prevent confusing bugs by zeroing out users provided data

#37 Add timeout so operations to the frame can fail if the frame fails to load for some reason

### CODE REVIEW

File: src/shim/sdk/DocumentSDK.ts [hunk 1]
```diff
@@ -129,7 +129,7 @@ export function decrypt(documentID: string, documentData: Uint8Array) {
         type: "DOCUMENT_DECRYPT",
         message: {
             documentID,
-            documentData,
+            documentData: documentData.slice(),
```
Comment (skeet70) [inline src/shim/sdk/DocumentSDK.ts]: https://jsperf.com/new-array-vs-splice-vs-slice/151

Looks like the right one! `slice(0)` is a bit faster for me on FF, but `slice()` is only barely slower. In Chrome `slice()` is significantly the fastest.

File: src/shim/sdk/DocumentSDK.ts [hunk 2]
```diff
@@ -392,7 +392,7 @@ export const advanced = {
         const payload: MT.DocumentUnmanagedEncryptRequest = {
             type: "DOCUMENT_UNMANAGED_ENCRYPT",
             message: {
-                documentData,
+                documentData: documentData.slice(0),
```
Comment (skeet70) [inline src/shim/sdk/DocumentSDK.ts]: Why the difference here?

File: src/shim/sdk/tests/DocumentSDK.test.ts [hunk 3]
```diff
@@ -246,35 +246,9 @@ describe("DocumentSDK", () => {
...
-                            message: {
-                                documentID: "mydoc",
-                                documentData: doc,
-                            },
-                        },
-                        [doc]
-                    );
+                    //Ensure we cloned these bytes before passing them to the frame
+                    const passedDoc = (FrameMediator.sendMessage as jasmine.Spy).calls.mostRecent().args[0].message.documentData;
+                    expect(passedDoc).not.toBe(doc);
```
Comment (skeet70) [inline src/shim/sdk/tests/DocumentSDK.test.ts]: Probably also should check `expect(passedDoc).toEqual(doc)` if that wasn't already covered.

File: src/shim/sdk/tests/DocumentSDK.test.ts [hunk 4]
```diff
@@ -319,6 +293,9 @@ describe("DocumentSDK", () => {
                         },
                         [document]
                     );
+                    //Ensure we cloned these bytes before passing them to the frame
+                    const passedDoc = (FrameMediator.sendMessage as jasmine.Spy).calls.mostRecent().args[0].message.documentData;
+                    expect(passedDoc).not.toBe(document);
```
Comment (skeet70) [inline src/shim/sdk/tests/DocumentSDK.test.ts]: Same here and the rest of the spots this is being checked, unless it's covered somewhere lower down/more specific.

File: src/shim/sdk/DocumentSDK.ts (hunk 2 above)
Comment (ernieturner) [reply src/shim/sdk/DocumentSDK.ts]: No reason, just left over from initial code. Changed.

File: src/shim/sdk/tests/DocumentSDK.test.ts (hunk 3 above)
Comment (ernieturner) [reply src/shim/sdk/tests/DocumentSDK.test.ts]: Ya, thats asserted about up above on line 239.

File: src/shim/sdk/tests/DocumentSDK.test.ts (hunk 4 above)
Comment (ernieturner) [reply src/shim/sdk/tests/DocumentSDK.test.ts]: Ditto from above, asserted about in the `.toHaveBeenCalledWith` assertion.

**Pull request merged**